The following exercise is an attempt to explore the Poisson probability distribution function as well as minor data clean-ups using pandas.

. . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 

<strong>Motivation for studying:</strong> I'm curious to know when the next big disaster caused by some tropical depression is about to happen in the Philippines.

<strong>Why the Poisson model?</strong> The Poisson distribution is also referred to as a limiting case of the Binomial model, which ascertains the probability of success or occurrence of an event, p, given a certain number of Bernoulli trials, n. Each "trial" may also refer to intervals of time, say years, hours or seconds. In this regard, in situations wherein the occurrence of the event in question is rare across many, many trials (or years, for example), using the Poisson model is easier. Thus, as n approaches infinity and as p becomes very small, we could use a different variable to denote the rate or number of occurrences of the desired event by multiplying n with p. More formally, this is the lambda used in the probability distribution function of the Poisson model.

Below is the PDF of the Poisson distribution:
$$p(x) = \frac{e^{-\lambda}\lambda^x}{x!}$$

To use the Poisson model, we need to make sure that the following assumptions hold true in this particular scenario we want to forecast:
1. All events must be independent of each other.
2. The probability of success or occurrence of an event is constant.
3. As the interval becomes smaller and smaller, the probability of success or occurrence becomes zero.

<strong>Definition of terms:</strong>
* Event: Super typhoon (as tagged by PAG-ASA)
* ${\lambda}$ = (number of trials/intervals, n) x (probability of occurrence within each trial/interval, v)
* random variable X = number of a super typhoon's occurrence
* p(X=x) = probability that x super typhoon/s will occur
* t = length of time interval

For this particular exercise, I will assume that the following conditions are true. Please note that I am not a meteorologist and if there is any particular assumption below that should be corrected, please let me know:
1. The occurrence or non-occurence of super typhoons in a year does not depend on a previous nor a possible future occurrence or non-occurrence.
2. The probability of having a super typhoon within the same interval is the constant.
3. The probability of a super typhoon occurring within a very short time interval is zero.

I know that typhoons in the Philippines tend to happen during the rainy season (latter half of the year) so I preferred using <u>number of years</u> as the time interval to avoid any bias. 

<strong>Data set sources: </strong>
<br>https://en.wikipedia.org/wiki/List_of_retired_Philippine_typhoon_names#cite_note-others-3 
<br>https://www.typhoon2000.ph/stormstats/WPF_StrongestTyphoonsPhilippines_2015Ed.pdf
<br><i>**can't find anything else more reliable</i>

In [2]:
import wikitablescrape

wikitablescrape.scrape(url="https://en.wikipedia.org/wiki/List_of_retired_Philippine_typhoon_names#cite_note-others-3",
                       output_name="ph_typhoons1")

In [2]:
import pandas as pd
df1 = pd.read_csv('ph_typhoons1.csv')
df2 = pd.read_csv('ph_typhoons1_1.csv')
df3 = pd.read_csv('ph_typhoons1_2.csv')

In [6]:
df = df1.append(df2.append(df3,sort=True),sort=True)

In [10]:
df1.columns

Index(['PAGASAName', 'WMOname', 'Replacementname', 'Dates active',
       'PAGASACategory', 'Sustainedwind speeds', 'Pressure', 'Areas affected',
       'Damage(PHP)', 'Deaths', 'Missing', 'Refs'],
      dtype='object')

In [11]:
df2.columns

Index(['LocalName', 'WMOname', 'Replacementname', 'Dates active', 'Category',
       'Wind', 'Pressure', 'Areas affected', 'Damage', 'Deaths', 'Missing',
       'Refs'],
      dtype='object')

In [12]:
df3.columns

Index(['LocalName', 'WMOname', 'Replacementname', 'Dates active(within PAR)',
       'PAGASACategory', 'Sustainedwind speeds', 'Pressure', 'Areas affected',
       'Damage(PHP)', 'Deaths', 'Missing', 'Refs'],
      dtype='object')

In [19]:
#renaming columns so everything appends nicely
df1.rename(columns={'PAGASAName':'LocalName'},inplace=True)
df2.rename(columns={'Category':'PAGASACategory','Wind':'Sustainedwind speeds','Damage':'Damage(PHP)'},inplace=True)
df3.rename(columns={'Dates active(within PAR)':'Dates active'},inplace=True)

In [68]:
df1.tail(2)

LocalName WMOname Replacementname           Dates active PAGASACategory  \
21    Loleng    Babs             NaN  October 15 – 24, 1998  Super Typhoon   
22  22 names     NaN   17.9 thousand                   4730            NaN   

   Sustainedwind speeds              Pressure  Areas affected   Damage(PHP)  \
21   155 km/h (100 mph)  940 hPa (27.38 inHg)  Visayas, Luzon  6.79 billion   
22                  NaN                   NaN             NaN           NaN   

   Deaths Missing  Refs  
21    303      29   NaN  
22    NaN     NaN   NaN

In [69]:
df2.tail(2)

LocalName WMOname Replacementname                     Dates active  \
10    Pepeng   Parma           Paolo  September 30 – October 10, 2009   
11  10 names     NaN   ₱66.6 billion                            >5024   

   PAGASACategory Sustainedwind speeds              Pressure  Areas affected  \
10        Typhoon   185 km/h (115 mph)  938 hPa (27.70 inHg)  Visayas, Luzon   
11           1814                  NaN                   NaN             NaN   

      Damage(PHP) Deaths  Missing  Refs  
10  ₱39.6 billion    465     47.0   NaN  
11            NaN    NaN      NaN   NaN

In [71]:
df3.tail(2)

LocalName WMOname Replacementname            Dates active  \
23    Rosita    Yutu           Rosal   October 27 – 31, 2018   
24     Usman    ----         Umberto  December 25 – 29, 2018   

         PAGASACategory Sustainedwind speeds               Pressure  \
23              Typhoon   215 km/h (130 mph)   900 hPa (26.58 inHg)   
24  Tropical Depression     55 km/h (35 mph)  1000 hPa (29.53 inHg)   

             Areas affected    Damage(PHP) Deaths Missing  Refs  
23            Central Luzon   ₱2.9 billion     27       0   NaN  
24  Visayas, Southern Luzon  ₱5.41 billion    155       0   NaN

In [73]:
# Both df1 and df2 have unneccessary summary rows at the end, so remove it
df1.drop(df1.index[-1], inplace=True)
df2.drop(df2.index[-1], inplace=True)

In [74]:
df1.tail(2)

LocalName WMOname Replacementname                   Dates active  \
19    Rosing  Angela          Rening  October 25 – November 7, 1995   
20    Iliang     Zeb             NaN             October 7–14, 1998   

   PAGASACategory Sustainedwind speeds              Pressure  Areas affected  \
19  Super Typhoon   215 km/h (130 mph)  910 hPa (26.87 inHg)  Southern Luzon   
20  Super Typhoon   205 km/h (125 mph)  900 hPa (26.58 inHg)  Southern Luzon   

     Damage(PHP) Deaths Missing  Refs  
19  10.8 billion    936     NaN   NaN  
20  5.38 billion     46      29   NaN

In [75]:
df2.tail(2)

LocalName  WMOname Replacementname                     Dates active  \
9      Ondoy  Ketsana          Odette          September 24 — 27, 2009   
10    Pepeng    Parma           Paolo  September 30 – October 10, 2009   

   PAGASACategory Sustainedwind speeds              Pressure  Areas affected  \
9         Typhoon    130 km/h (80 mph)  980 hPa (28.94 inHg)           Luzon   
10        Typhoon   185 km/h (115 mph)  938 hPa (27.70 inHg)  Visayas, Luzon   

      Damage(PHP) Deaths  Missing  Refs  
9   ₱11.2 billion    671     37.0   NaN  
10  ₱39.6 billion    465     47.0   NaN

In [76]:
df = df1.append(df2.append(df3,sort=True),sort=True)

In [77]:
df.drop(['Refs','WMOname','Replacementname'],axis=1,inplace=True)

In [78]:
df.set_index('LocalName',inplace=True)

In [79]:
df.head()

Areas affected  Damage(PHP)  \
LocalName                                                                  
Dading                                        Central Luzon      Unknown   
Welming    Eastern, Northeastern Visayas and Southern Luzon          NaN   
Pitang                                       Northern Luzon  1.4 million   
Sening                 Southern Luzon, Northeastern Visayas   74 million   
Titang                            Mindanao, Western Visayas   50 million   

                            Dates active Deaths Missing PAGASACategory  \
LocalName                                                                
Dading            June 26 - July 3, 1964    100     NaN        Typhoon   
Welming    October 31 - November 8, 1967    300      64  Super Typhoon   
Pitang              September 8–14, 1970     95      80  Super Typhoon   
Sening               October 10–18, 1970    768     193  Super Typhoon   
Titang               October 14–25, 1970  1,551     284  Super Typhoon   

                       Pressure Sustainedwind speeds  
LocalName                                             
Dading     970 hPa (28.64 inHg)   185 km/h (115 mph)  
Welming    910 hPa (26.87 inHg)   260 km/h (160 mph)  
Pitang     905 hPa (26.72 inHg)   260 km/h (160 mph)  
Sening     905 hPa (26.72 inHg)   280 km/h (175 mph)  
Titang     940 hPa (27.76 inHg)   240 km/h (150 mph)

In [80]:
#need to create the following columns for easier computations:
df['Year Active'] = df['Dates active'].str[-4:]

In [81]:
df.tail()

Areas affected    Damage(PHP)             Dates active  \
LocalName                                                                    
Urduja                     Visayas  ₱3.75 billion   December 11 – 19, 2017   
Vinta                     Mindanao   ₱2.1 billion   December 20 – 24, 2017   
Ompong                       Luzon  ₱33.9 billion  September 12 – 15, 2018   
Rosita               Central Luzon   ₱2.9 billion    October 27 – 31, 2018   
Usman      Visayas, Southern Luzon  ₱5.41 billion   December 25 – 29, 2018   

          Deaths Missing       PAGASACategory               Pressure  \
LocalName                                                              
Urduja        54       0       Tropical Storm   996 hPa (29.41 inHg)   
Vinta        173     176              Typhoon   970 hPa (28.64 inHg)   
Ompong       127       0        Super Typhoon   905 hPa (26.72 inHg)   
Rosita        27       0              Typhoon   900 hPa (26.58 inHg)   
Usman        155       0  Tropical Depression  1000 hPa (29.53 inHg)   

          Sustainedwind speeds Year Active  
LocalName                                   
Urduja        75 km/h (45 mph)        2017  
Vinta        130 km/h (80 mph)        2017  
Ompong      205 km/h (125 mph)        2018  
Rosita      215 km/h (130 mph)        2018  
Usman         55 km/h (35 mph)        2018

In [82]:
type(df['Year Active'][0])

str

In [83]:
#convert to int
df['Year Active'] = df['Year Active'].apply(pd.to_numeric)

In [84]:
type(df['Year Active'][0])

numpy.int64

<strong>Problem I-A: What is the probability that there is no super typhoon in one year?</strong>

In [108]:
import numpy as np

In [109]:
#find the probability or rate of super typhoon occurrence per year:
v = df.PAGASACategory.value_counts()['Super Typhoon']/abs(df['Year Active'].max()-df['Year Active'].min())
n_a = 1
x = 0

In [110]:
v

0.2777777777777778

This means that based on historical data since 1964, we have experienced about 0.2778 super typhoons per year.

In [111]:
lambda_ = v*n_a

In [114]:
p_a = ((np.exp(-lambda_))*(lambda_**x))/np.math.factorial(x)
p_a

0.7574651283969664

Thus, the probability of <u>not experiencing</u> a super typhoon in one year is about 75.75%. Which also means that the probability of experiencing one is about 24.25%. The odds of no super typhoon is approximately 3x against having a super typhoon.

<strong>Problem I-B: How about within the next two-and-a-half years?</strong>

In [118]:
n_b = 2.5
lambda_b = v*n_b

In [119]:
p_b = ((np.exp(-lambda_b))*(lambda_b**x))/np.math.factorial(x)
p_b

0.4993517885992762

Here, we can see that the probability of not experiencing vs experiencing a super typhoon gets close to almost 50:50.

<strong>Problem I-C: How about within the next five years?</strong>

In [131]:
n_c = 5
lambda_c = v*n_c

In [132]:
p_c = ((np.exp(-lambda_c))*(lambda_c**x))/np.math.factorial(x)
p_c

0.24935220877729622

The probability of not experiencing a super typhoon within the next five years goes down to just about 24.93%, while raising the probability of having a super typhoon to about 75.07%

<strong>Problem II: What is probability of having three super typhoons in the next two years?</strong>

In this particular question, seeing the Poisson model as a limiting case of the Binomial model might no longer be accurate since we are no longer only looking at whether an event occurs or does not occur (1 or 0). However, we could still use the Poisson distribution to ascertain the probability of the event in question.

In [123]:
n_ii = 2
x_ii = 3

In [124]:
lambda_ii = v*n_ii

In [126]:
p_ii = ((np.exp(-lambda_ii))*(lambda_ii**x_ii))/np.math.factorial(x_ii)
p_ii

0.016396702695971446

Hence, the probability of experiencing three super typhoons within the next two years is about 1.64%.

<h3>Final words</h3>

How reliable are these probabilities? It will only be reliable if and only if the assumptions I have stated above are correct, and that no underlying conditions (like climate change) make the occurrences of super typhoons to be more frequent.

. . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 

<font size='3'>This post is heavily inspired by the following blogs:
<br>- https://medium.com/@ns2586/using-poisson-distribution-to-forecast-the-next-earthquake-38a4406ab71
<br>- https://probabilityandstats.wordpress.com/2011/08/18/poisson-as-a-limiting-case-of-binomial-distribution/</font>